Add scripts/kwreview.sh

joaoseckler opened the pull request:
This script aids kw's developers in running shfmt and shellcheck. It
uses reviewdog in order to check only for warnings and errors in the
curent patch (compared to branch unstable by default). It also updates
the "Development workflow" documentation to show the usage of this
script.

Signed-off-by: João Seckler <jseckler@riseup.net>

alan-barzilay commented on documentation/content/developmentworkflow.rst:
```diff
@@ -120,3 +120,29 @@ To configure this Git pre-commit hook:
+Kwreview
+========
+Another way to have your code checked against ``shfmt`` and ``shellcheck`` is to
+use ``scripts/kwreview.sh``. By default, it will check you current patch (i.e. what
```
your* current patch

rodrigosiqueira commented:
Hi @joaoseckler ,
This PR looks great!
By the way, I would like to see an explanation of how to add this to git hooks, e.g., `.git/hooks/pre-commit.sh`. Also, could you organize your code in a way that uses functions and better describes each part's intention?

joaoseckler commented:
The last update of this PR organizes the code into functions.

joaoseckler commented:
> This PR looks great!

Thanks!

> By the way, I would like to see an explanation of how to add this to git hooks, e.g., `.git/hooks/pre-commit.sh.`

I've come up with this hook. Do you think it's worth to adding it to the documentation?
```bash
#!/bin/bash

output="$(./scripts/kwreview.sh)"
if [ "$output" != "" ]; then
  echo "$output"
  echo
  echo "Please fix the errors above before committing."
  exit 1
else
  exit 0
fi
```

joaoseckler commented:
Future work on this PR includes:
 - Documenting functions
 - Using appropriate returnr values (instead of only 1 and 0)

rodrigosiqueira commented on documentation/content/developmentworkflow.rst:
```diff
@@ -120,3 +120,29 @@ To configure this Git pre-commit hook:
+
+Now you can call ``kwreview.sh`` with the vim command ``:Kwreview``
+(notice the upper case initial) and navigate the errors/warnings with
+``:copen``, ``:cn``, ``:cp``, etc.
```
It looks like that your script depends on reviewdog. You need to highlight this information and tell how to install it on Debian and Arch.

rodrigosiqueira replied:
Maybe we can add a new option under `setup -i` where we install development dependencies.

rodrigosiqueira commented on scripts/kwreview.sh:
```diff
@@ -0,0 +1,214 @@
+  opt['branch']='unstable'
+  opt['path']='.'
+  opt['exclude']=''
+  opt['shellcheck_exclude_default']='SC2016,SC2181,SC2034,SC2154'
```
how about reading this from the `.travis` file?

rodrigosiqueira commented on scripts/kwreview.sh:
```diff
@@ -0,0 +1,214 @@
+    -- "$@")
+
+  # shellcheck disable=2086
+  # see man getopt
```
which section?

rodrigosiqueira commented on scripts/kwreview.sh:
```diff
@@ -0,0 +1,214 @@
+  # see man getopt
+  eval set --$options
+
+  while [[ $# -gt 0 ]]; do
```
**"**#**"**

rodrigosiqueira commented on scripts/kwreview.sh:
```diff
@@ -0,0 +1,214 @@
+        shift 2
+        ;;
+      --all | -a)
+        opt['shellcheck_exclude']=""
```
'' instead of ""

rodrigosiqueira commented on scripts/kwreview.sh:
```diff
@@ -0,0 +1,214 @@
+  local shfmt_inplace="${opt['shfmt_inplace']}"
+  local branch="${opt['branch']}"
+
+  if [ "$shellcheck_only" = 'FALSE' ]; then
```
can we usee `[[` and `==` here?

rodrigosiqueira reviewed (changes requested):
In general, it looks really good. However, it lacks some information on how to make it work, and I also felt that you used some different code styles here (especially in the `if` conditions).

joaoseckler commented:
The last update to this PR improves the documentation, accepts siqueira's suggestions and makes use of better exit codes.

rodrigosiqueira commented on documentation/content/developmentworkflow.rst:
```diff
@@ -120,3 +120,74 @@ To configure this Git pre-commit hook:
+Installing pre-commit
+~~~~~~~~~~~~~~~~~~~~~
+``pre-commit`` is available as a package for most distributions. On debian based
+distros, use ``sudo apt install pre-commit``. On Arch based distros, use ``sudo
```
It looks like that we don't have this pre-commit on Ubuntu yet :(
Do you know how I can work around this?

alan-barzilay replied:
@rodrigosiqueira `pip install pre-commit`

rodrigosiqueira commented on documentation/content/developmentworkflow.rst:
```diff
@@ -120,3 +120,74 @@ To configure this Git pre-commit hook:
+~~~~~~~~~~~~~~~~~~~~
+To install reviewdog, run the following command::
+
+  curl -sfL https://raw.githubusercontent.com/reviewdog/reviewdog/master/install.sh | sh -s
```
fwiu, if you use this method, it will install reviewdog in `$PWD/bin/`. It looks like that your script does not handle this situation. Also, even after changing your code, it does not give me any update; I tried in the `src/plugins/subsystems/drm/drm.sh` which has some shellcheck problems, but I did not get anything (:Kwreview)

joaoseckler replied:
> fwiu, if you use this method, it will install reviewdog in $PWD/bin/. It looks like that your script does not handle this situation.

I changed the documentation a bit, see if it is better. 

> Also, even after changing your code, it does not give me any update; I tried in the src/plugins/subsystems/drm/drm.sh which has some shellcheck problems, but I did not get anything (:Kwreview)

By default, `kwreview` will compare the current patch with unstable, and will ignore the staging area and the working tree. I just added an option to compare against working tree (`-t`) and agains the staging area (`-g`). Tell me if that was not the problem.

rodrigosiqueira commented:
@joaoseckler did you update this PR?

joaoseckler commented:
> @joaoseckler did you update this PR?

I had unpushed changes in my local repository... Now they are here. They add the `-g` and `-t` flag and the suggestion to install pre-commit with pip instead of apt or pacman.

joaoseckler commented:
The last update to this PR adds the `-f` / `--file` option, which allows for a direct checking of files, without checking the VCS.

joaoseckler commented:
> The last update to this PR adds the `-f` / `--file` option, which allows for a direct checking of files, without checking the VCS.

On a second thought, this is better implemented with an `--ignore-vcs`/`-i` option, which considers all files passed by the user in the command line. We also now make use of the `-filter-mode` flag of reviewdog, making the code cleaner. Since we were dealing with it, we also implemented a `--filter-mode` option to kwreview. We also fixed a problem with `--shfmt-only`, which was not working.

magalilemes reviewed (approved):
Reaaally cool work.

Only thing I'd check is that you wrote in developmentworkflow.rst that pre-commit is available on debian-based distros, but I think it's still in the experimental phase. Besides that, everything looks great and easy to use.

alan-barzilay reviewed (approved):
lgtm, nice job!

rodrigosiqueira commented:
Hi @joaoseckler ,
This new change is great; I really like the `-i` option. However, I don't think the name `--ignore-vcs` is intuitive (actually, what vcs stands for?). I have to admit that I prefer something like `--file`... how about make `./script/kreview.sh PATH` do the same work as `-i` does? Also, add the instruction for this option in the `developmentworkflow.rst` file.

alan-barzilay commented:
@rodrigosiqueira vcs stands for version control system

joaoseckler commented:
The last update to this PR accepts the suggestion of using the -i option by default, changes the documentation about that, and change the documentation regarding the installation of pre-commit.

rodrigosiqueira commented:
Hi @joaoseckler ,
Thanks a lot for all your effort here, merged to the unstable branch.
